### PR TITLE
Support extending revset functionality with custom functions and programmatic filters

### PR DESCRIPTION
**Pull request by torquestomp**
This doesn't support _everything_ an extension might want to do with the revset language, but it covers a large portion of it and for much less internal cost than the previous proposal.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/revset.rs [hunk 1]
```diff
@@ -1152,7 +1155,7 @@ fn parse_function_expression(
         state.with_alias_expanding(id, &locals, primary_span, |state| {
             parse_program(defn, state)
         })
-    } else if let Some(func) = BUILTIN_FUNCTION_MAP.get(name) {
+    } else if let Some(func) = state.function_map.get(name) {
```
Comment (jonathantanmy) [inline lib/src/revset.rs:1180]: Do we need to make a similar change to `all_function_names` (to use state.function_map instead of BUILTIN_FUNCTION_MAP)?

**Comment by jonathantanmy** [comment]
Also is there a way to test the function and filter extensibility? I see that there is a CLI example that shows how to create a new filter, but that doesn't seem to cover the function ability, and does not seem to be exercised in a test.

**Comment by torquestomp** [comment]
> Also is there a way to test the function and filter extensibility? I see that there is a CLI example that shows how to create a new filter, but that doesn't seem to cover the function ability, and does not seem to be exercised in a test.

Unfortunately I don't think there's currently a good way to set up tests for extension behavior; the low level revset_engine tests do not have enough context to produce a `Store`, and the high-level tests run `jj` directly which does not have extensions loaded in.  Maybe we need some way to run `example` binaries instead of the main binary in tests

File: lib/src/revset.rs (hunk 1 above)
Comment (torquestomp) [reply lib/src/revset.rs:1180]: Yes, good catch.

File: lib/src/revset.rs [hunk 2]
```diff
@@ -2656,10 +2656,38 @@ impl RevsetExtensions {
...
+            extensions,
+            workspace,
+        }
+    }
+
+    pub fn aliases_map(&self) -> &'a RevsetAliasesMap {
+        self.aliases_map
+    }
+
+    pub fn user_email(&self) -> &String {
```
Comment (yuja) [inline lib/src/revset.rs:2737]: nit: could return `&str` and let caller do `.to_owned()`.

File: lib/src/revset.rs [hunk 3]
```diff
@@ -1152,22 +1155,28 @@ fn parse_function_expression(
...
+                    all_function_names(state.function_map.keys().copied(), state.aliases_map),
+                ),
             },
             name_pair.as_span(),
         ))
     }
 }
 
-fn all_function_names(aliases_map: &RevsetAliasesMap) -> impl Iterator<Item = &str> {
+fn all_function_names<'a>(
```
Comment (yuja) [inline lib/src/revset.rs:1196]: nit: inline this function (because it's now equivalent to `chain()`)?

File: lib/src/revset.rs [hunk 4]
```diff
@@ -2660,6 +2678,7 @@ pub struct RevsetParseContext<'a> {
     user_email: String,
     extensions: &'a RevsetExtensions,
     workspace: Option<RevsetWorkspaceContext<'a>>,
+    function_map: HashMap<&'static str, &'a RevsetFunction>,
```
Comment (yuja) [inline lib/src/revset.rs:2703]: nit: unless we plan to turn `RevsetFunction` into `Box<dyn Fn>`, `RevsetFunction` can be just copied.

File: lib/src/revset.rs [hunk 5]
```diff
@@ -2669,11 +2688,23 @@ impl<'a> RevsetParseContext<'a> {
...
     ) -> Self {
+        let mut function_map = HashMap::<&'static str, &'a RevsetFunction>::new();
+        for (name, func) in BUILTIN_FUNCTION_MAP.iter() {
+            function_map.insert(name, func);
+        }
+        for (name, func) in &extensions.custom_functions {
+            match function_map.entry(name) {
+                hash_map::Entry::Occupied(_) => panic!("Cannot override builtin function '{name}'"),
+                hash_map::Entry::Vacant(v) => v.insert(func),
+            };
```
Comment (yuja) [inline lib/src/revset.rs:2721]: nit: maybe better to detect name conflicts at `add_custom_extension()` so that bad extension can be blamed based on call stack.

File: Cargo.lock
Comment (yuja) [inline Cargo.lock:1]: nit: unrelated change

File: lib/src/revset.rs [hunk 6]
```diff
@@ -322,6 +323,25 @@ pub enum RevsetCommitRef {
...
 
+/// A custom revset filter expression, defined by an extension.
+pub trait RevsetFilterExtension: std::fmt::Debug + Any {
+    fn as_any(&self) -> &dyn Any;
+
+    /// Returns true iff this filter matches the specified commit.
+    fn matches_commit(&self, commit: &Commit) -> bool;
+
+    /// Returns true iff this filter is functionally equivalent to `ext`.
+    fn eq_ext(&self, ext: &dyn RevsetFilterExtension) -> bool;
```
Comment (yuja) [inline lib/src/revset.rs:334]: Perhaps, we can remove `Eq, PartialEq` from `Expression` enums. I don't think all future core filter expressions do support Eq.

I'll take a look.

File: lib/src/revset.rs [hunk 7]
```diff
@@ -336,6 +356,8 @@ pub enum RevsetFilterPredicate {
     File(FilesetExpression),
     /// Commits with conflicts
     HasConflict,
+    /// Custom predicates provided by extensions
+    Extension(Rc<Box<dyn RevsetFilterExtension>>),
```
Comment (yuja) [inline lib/src/revset.rs:360]: nit: and `Box<` can probably be removed if `Eq` bound is removed.

**Review by yuja — APPROVED** [review]
Looks generally good to me, thanks!

File: lib/src/revset.rs (hunk 6 above)
Comment (yuja) [reply lib/src/revset.rs:334]: Hmm, it seems better to rewrite the parser first so that parsing tests don't depend on equivalence of `RevsetExpression` trees. Until then, maybe we can use pointer-based equivalence? I assume the backend doesn't need `eq_ext()`, and will use `as_any()` to extract the underlying representation.

https://doc.rust-lang.org/std/rc/struct.Rc.html#method.ptr_eq
